### Image Recognition using a Convolutional Neural Network (CNN) classification model for the CIFAR-10 dataset.

Data Source:
http://www.cs.toronto.edu/~kriz/cifar.html


I used Google Colab for this analysis. The personal computer CPUs can not handle the computation required to train a CNN model. Colab, however is equipped with GPUs that are optimised for these tasks.

In [2]:
# Installing the required libraries:

!pip install tensorflow numpy

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [3]:
# Importing the necessary libraries:

import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import cifar10
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Dense, Flatten, Dropout, BatchNormalization
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.utils import to_categorical


In [4]:
# Loading and preprocessing the CIFAR-10 dataset:

# Load the CIFAR-10 dataset
(train_images, train_labels), (test_images, test_labels) = cifar10.load_data()

# Normalize the pixel values to be between 0 and 1
train_images = train_images / 255.0
test_images = test_images / 255.0

# Convert the labels to one-hot encoded vectors
train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)


170498071/170498071 [==============================] - 2s 0us/step


In [5]:
# Defining the CNN model:

# Create a sequential model
model = Sequential()

# Add a convolutional layer with 32 filters, a 3x3 kernel, ReLU activation, and input shape
model.add(Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
model.add(BatchNormalization())

# Add a convolutional layer with 32 filters, a 3x3 kernel, and ReLU activation
model.add(Conv2D(32, (3, 3), activation='relu'))
model.add(BatchNormalization())

# Add a max-pooling layer with a 2x2 pool size and 25% dropout
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

# Add a convolutional layer with 64 filters, a 3x3 kernel, and ReLU activation
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(BatchNormalization())

# Add a convolutional layer with 64 filters, a 3x3 kernel, and ReLU activation
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(BatchNormalization())

# Add a max-pooling layer with a 2x2 pool size and 25% dropout
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

# Add a flatten layer to convert 3D feature maps to 1D feature vectors
model.add(Flatten())

# Add a dense layer with 512 units and ReLU activation
model.add(Dense(512, activation='relu'))
model.add(BatchNormalization())
model.add(Dropout(0.5))

# Add a dense output layer with 10 units (for 10 classes) and a softmax activation
model.add(Dense(10, activation='softmax'))


In [6]:
# Compiling the model:

model.compile(optimizer=Adam(learning_rate=0.001),
              loss='categorical_crossentropy', metrics=['accuracy'])


In [7]:
# Training the model:

model.fit(train_images, train_labels, epochs=50,
          batch_size=64, validation_split=0.1)


Epoch 1/50
704/704 [==============================] - 232s 325ms/step - loss: 1.6417 - accuracy: 0.4611 - val_loss: 1.3143 - val_accuracy: 0.5366
Epoch 2/50
704/704 [==============================] - 231s 329ms/step - loss: 1.0940 - accuracy: 0.6164 - val_loss: 0.9686 - val_accuracy: 0.6612
Epoch 3/50
704/704 [==============================] - 230s 326ms/step - loss: 0.9407 - accuracy: 0.6702 - val_loss: 0.8442 - val_accuracy: 0.7056
Epoch 4/50
704/704 [==============================] - 229s 326ms/step - loss: 0.8523 - accuracy: 0.6983 - val_loss: 0.7455 - val_accuracy: 0.7468
Epoch 5/50
704/704 [==============================] - 229s 325ms/step - loss: 0.7842 - accuracy: 0.7241 - val_loss: 0.9806 - val_accuracy: 0.6724
Epoch 6/50
704/704 [==============================] - 228s 324ms/step - loss: 0.7352 - accuracy: 0.7425 - val_loss: 0.8144 - val_accuracy: 0.7248
Epoch 7/50
704/704 [==============================] - 229s 326ms/step - loss: 0.6929 - accuracy: 0.7545 - val_loss: 0.6932 -

In [8]:
# Evaluating the model:
# We evaluate the model on the test dataset and print the test accuracy.


test_loss, test_accuracy = model.evaluate(test_images, test_labels)
print(f'Test accuracy: {test_accuracy:.4f}')

313/313 [==============================] - 12s 39ms/step - loss: 0.5983 - accuracy: 0.8246
Test accuracy: 0.8246


This CNN-based classification model for the CIFAR-10 dataset achieves a test accuracy of 82%, which can be further improved by adding more layers, using data augmentation, adjusting the hyperparameters (like epoch number of 100), or employing more advanced architectures like Residual Networks (ResNet).